File: src/main/java/hu/frontrider/arcana/creatureenchant/backend/CreatureEnchant.java
package hu.frontrider.arcana.creatureenchant.backend;

import hu.frontrider.arcana.capabilities.CreatureEnchantCapability;
import hu.frontrider.arcana.capabilities.ICreatureEnchant;
import hu.frontrider.arcana.util.AspectUtil;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.fml.common.registry.GameRegistry;
import net.minecraftforge.registries.IForgeRegistryEntry;
import thaumcraft.api.aspects.AspectList;

import java.util.Collections;
import java.util.Map;

import static hu.frontrider.arcana.ThaumicArcana.MODID;
import static hu.frontrider.arcana.capabilities.CreatureEnchantProvider.CREATURE_ENCHANT_CAPABILITY;

public abstract class CreatureEnchant extends IForgeRegistryEntry.Impl<CreatureEnchant> {

    @GameRegistry.ObjectHolder(MODID +":normal")
    private static EnchantingBaseCircle baseCircle = null;

    private final ResourceLocation icon;
    private final String unlocalizedName;

    public CreatureEnchant(ResourceLocation resourceLocation, String unlocalizedName) {
        setRegistryName(resourceLocation);
        this.unlocalizedName = unlocalizedName;
        icon = new ResourceLocation(resourceLocation.getResourceDomain(), "textures/cenchant/" + resourceLocation.getResourcePath() + ".png");
    }

    public String getUnlocalizedName() {
        return "enchant.creature_enchant." + unlocalizedName;
    }

    public int getEnchantLevel(EntityLivingBase entity, CreatureEnchant enchantment) {
        if (entity.hasCapability(CREATURE_ENCHANT_CAPABILITY, null)) {
            ICreatureEnchant capability = entity.getCapability(CREATURE_ENCHANT_CAPABILITY, null);
            int level = capability.getLevel(enchantment);
            if (capability.hasEnchant(enchantment))
            return capability.getCircle().doEffect(level,entity,enchantment);

        }
        return 0;
    }

    public static void setEnchantment(Entity entity,ICreatureEnchant enchants) {
        if (entity.hasCapability(CREATURE_ENCHANT_CAPABILITY, null)) {
            ICreatureEnchant capability = entity.getCapability(CREATURE_ENCHANT_CAPABILITY, null);

            capability.setStore(enchants.getStore());
            capability.setCircle(enchants.getCircle());
        }
    }

    public static boolean isEnchanted(Entity entity) {
        if (entity.hasCapability(CREATURE_ENCHANT_CAPABILITY, null)) {
            ICreatureEnchant capability = entity.getCapability(CREATURE_ENCHANT_CAPABILITY, null);
            return capability.hasEnchant();
        }
        return false;
    }

    public static Map<CreatureEnchant, CreatureEnchantCapability.CreatureEnchantContainer> getCreatureEnchants(Entity entity) {
        if (entity.hasCapability(CREATURE_ENCHANT_CAPABILITY, null)) {
            ICreatureEnchant capability = entity.getCapability(CREATURE_ENCHANT_CAPABILITY, null);
            return capability.getStore();
        }

        return Collections.emptyMap();
    }

    public static EnchantingBaseCircle getBaseCircle(Entity entity){
        if (entity.hasCapability(CREATURE_ENCHANT_CAPABILITY, null)) {
            ICreatureEnchant capability = entity.getCapability(CREATURE_ENCHANT_CAPABILITY, null);
            return capability.getCircle();
        }

        return baseCircle;
    }


    public static CreatureEnchant getForFormula(AspectList formula){
        for (CreatureEnchant enchant : GameRegistry.findRegistry(CreatureEnchant.class).getValuesCollection()) {
            if (AspectUtil.INSTANCE.aspectListEquals(enchant.formula(), formula))
                return enchant;
        }
        return null;
    }

    public abstract AspectList formula();

    public abstract String getResearch();


    @Override
    public String toString() {
        return "CreatureEnchant{" +
                ", registryName=" + getRegistryName() +
                ", research='" + getResearch() + '\'' +
                '}';
    }

    public ResourceLocation getIcon() {
        return icon;
    }
}

File: src/main/java/hu/frontrider/arcana/creatureenchant/CreatureEnchant.kt
package hu.frontrider.arcana.creatureenchant

import hu.frontrider.arcana.capabilities.creatureenchant.ICreatureEnchant
import hu.frontrider.arcana.util.AspectUtil
import net.minecraft.entity.Entity
import net.minecraft.entity.EntityLivingBase
import net.minecraft.util.ResourceLocation
import net.minecraftforge.fml.common.registry.GameRegistry
import net.minecraftforge.registries.IForgeRegistryEntry
import thaumcraft.api.aspects.AspectList

import hu.frontrider.arcana.ThaumicArcana.MODID
import hu.frontrider.arcana.capabilities.creatureenchant.CreatureEnchantCapability
import hu.frontrider.arcana.capabilities.creatureenchant.CreatureEnchantProvider.Companion.CREATURE_ENCHANT_CAPABILITY

abstract class CreatureEnchant(resourceLocation: ResourceLocation, private val unlocalizedName: String) : IForgeRegistryEntry.Impl<CreatureEnchant>() {

    val icon: ResourceLocation

    abstract val research: String

    init {
        registryName = resourceLocation
        icon = ResourceLocation(resourceLocation.resourceDomain, "textures/cenchant/" + resourceLocation.resourcePath + ".png")
    }

    fun getUnlocalizedName(): String {
        return "enchant.creature_enchant.$unlocalizedName"
    }

    open fun getEnchantLevel(entity: EntityLivingBase, enchantment: CreatureEnchant): Int {
        if (entity.hasCapability(CREATURE_ENCHANT_CAPABILITY, null)) {
            val capability = entity.getCapability(CREATURE_ENCHANT_CAPABILITY, null)!!
            val level = capability.getLevel(enchantment)
            if (capability.hasEnchant(enchantment))
                return capability.circle.doEffect(level, entity, enchantment)
        }
        return 0
    }

    abstract fun formula(): AspectList


    override fun toString(): String {
        return "CreatureEnchant{" +
                ", registryName=" + registryName +
                ", research='" + research + '\''.toString() +
                '}'.toString()
    }

    companion object {

        @GameRegistry.ObjectHolder("$MODID:normal")
        private val baseCircle: EnchantingBaseCircle? = null

        fun setEnchantment(entity: Entity, enchants: ICreatureEnchant) {
            if (entity.hasCapability(CREATURE_ENCHANT_CAPABILITY, null)) {
                val capability = entity.getCapability(CREATURE_ENCHANT_CAPABILITY, null)

                capability!!.store = enchants.store
                capability.circle = enchants.circle
            }
        }

        fun isEnchanted(entity: Entity): Boolean {
            if (entity.hasCapability(CREATURE_ENCHANT_CAPABILITY, null)) {
                val capability = entity.getCapability(CREATURE_ENCHANT_CAPABILITY, null)
                return capability!!.hasEnchant()
            }
            return false
        }

        fun getCreatureEnchants(entity: Entity): Map<CreatureEnchant, CreatureEnchantCapability.CreatureEnchantContainer> {
            if (entity.hasCapability(CREATURE_ENCHANT_CAPABILITY, null)) {
                val capability = entity.getCapability(CREATURE_ENCHANT_CAPABILITY, null)
                return capability!!.store
            }

            return emptyMap()
        }

        fun getBaseCircle(entity: Entity): EnchantingBaseCircle? {
            if (entity.hasCapability(CREATURE_ENCHANT_CAPABILITY, null)) {
                val capability = entity.getCapability(CREATURE_ENCHANT_CAPABILITY, null)
                return capability!!.circle
            }

            return baseCircle
        }


        fun getForFormula(formula: AspectList): CreatureEnchant? {
            for (enchant in GameRegistry.findRegistry(CreatureEnchant::class.java).valuesCollection) {
                if (AspectUtil.aspectListEquals(enchant.formula(), formula))
                    return enchant
            }
            return null
        }
    }
}

